crates/red_knot_module_resolver/src/path.rs

```diff
@@ -779,343 +415,398 @@ impl ModuleSearchPath {
+                SearchPathInner::StandardLibraryCustom(search_path),
+                FilePath::System(absolute_path),
+            ) => absolute_path
+                .strip_prefix(search_path)
+                .ok()
+                .map(|relative_path| ModulePath {
+                    search_path: self.clone(),
+                    relative_path: relative_path.as_utf8_path().to_path_buf(),
+                }),
+            (SearchPathInner::StandardLibraryCustom(_), FilePath::Vendored(_)) => None,
+            (
+                SearchPathInner::StandardLibraryVendored(search_path),
+                FilePath::Vendored(absolute_path),
+            ) => absolute_path
+                .strip_prefix(search_path)
+                .ok()
+                .map(|relative_path| ModulePath {
+                    search_path: self.clone(),
+                    relative_path: relative_path.as_utf8_path().to_path_buf(),
+                }),
+            (SearchPathInner::StandardLibraryVendored(_), FilePath::System(_)) => None,
+            (SearchPathInner::SitePackages(search_path), FilePath::System(absolute_path)) => {
+                absolute_path
+                    .strip_prefix(search_path)
+                    .ok()
+                    .map(|relative_path| ModulePath {
+                        search_path: self.clone(),
+                        relative_path: relative_path.as_utf8_path().to_path_buf(),
+                    })
+            }
+            (SearchPathInner::SitePackages(_), FilePath::Vendored(_)) => None,
+            (SearchPathInner::Editable(search_path), FilePath::System(absolute_path)) => {
+                absolute_path
+                    .strip_prefix(search_path)
+                    .ok()
+                    .map(|relative_path| ModulePath {
+                        search_path: self.clone(),
+                        relative_path: relative_path.as_utf8_path().to_path_buf(),
+                    })
+            }
+            (SearchPathInner::Editable(_), FilePath::Vendored(_)) => None,
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn as_system_path_buf(&self) -> Option<&SystemPathBuf> {
+        match &*self.0 {
+            SearchPathInner::Extra(path)
+            | SearchPathInner::FirstParty(path)
+            | SearchPathInner::StandardLibraryCustom(path)
+            | SearchPathInner::SitePackages(path)
+            | SearchPathInner::Editable(path) => Some(path),
+            SearchPathInner::StandardLibraryVendored(_) => None,
+        }
+    }
+
+    #[must_use]
+    pub(crate) fn as_vendored_path_buf(&self) -> Option<&VendoredPathBuf> {
+        match &*self.0 {
+            SearchPathInner::StandardLibraryVendored(path) => Some(path),
+            SearchPathInner::Extra(_)
+            | SearchPathInner::FirstParty(_)
+            | SearchPathInner::StandardLibraryCustom(_)
+            | SearchPathInner::SitePackages(_)
+            | SearchPathInner::Editable(_) => None,
+        }
     }
 }
```

Comment:
We should return a `&SystemPath` and `&VendoredPath` instead of the `buf` version
```suggestion
    pub(crate) fn as_system_path_buf(&self) -> Option<&SystemPath> {
        match &*self.0 {
            SearchPathInner::Extra(path)
            | SearchPathInner::FirstParty(path)
            | SearchPathInner::StandardLibraryCustom(path)
            | SearchPathInner::SitePackages(path)
            | SearchPathInner::Editable(path) => Some(path),
            SearchPathInner::StandardLibraryVendored(_) => None,
        }
    }

    #[must_use]
    pub(crate) fn as_vendored_path_buf(&self) -> Option<&VendoredPath> {
        match &*self.0 {
            SearchPathInner::StandardLibraryVendored(path) => Some(path),
            SearchPathInner::Extra(_)
            | SearchPathInner::FirstParty(_)
            | SearchPathInner::StandardLibraryCustom(_)
            | SearchPathInner::SitePackages(_)
            | SearchPathInner::Editable(_) => None,
        }
    }
}
```

Reply:
> We should return a `&SystemPath` and `&VendoredPath` instead of the `buf` version

I don't mind making the change, but what's the advantage?

Reply:
It's the same as `&str` and `String`. It's preferred to take or return a `&str` because it is the more generic type. I think might even avoid an extra dereference because a `&SystemPathBuf` is a reference to a `SystemPathBuf` and you then need to dereference the value. `&SystemPath` is directly the reference to the value.